docs/proposals/parameterized-config-management-plugins.md

```diff
@@ -0,0 +1,477 @@
+        cat << EOM
+        {
+          "kind": "ConfigMap",
+          "apiVersion": "v1",
+          "metadata": {
+            "name": "$ARGOCD_APP_NAME-$CM_NAME_SUFFIX",
+            "namespace": "$ARGOCD_APP_NAMESPACE"
+          }
+        }
+        EOM
+  discover:
+    fileName: "./trivial-cmp"
+  parameters:
+    command:
+      - sh
+      - -c
+      - |
+        echo '[{"name": "cm-name-suffix"}]'
+```
+
+#### Example 2: Helm parameters from Kustomize dependency
+
+**Plugin config**
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: ConfigManagementPlugin
+metadata:
+  name: kustomize-helm-proxy-cmp
+spec:
+  version: v1.0
+  generate:
+    command: [kustomize, build, ".", --enable-helm]
+  discover:
+    fileName: "./kustomization.yaml"
+  parameters:
+    command: [/home/argocd/get-parameters.sh]
+```
+
+**get-parameters.sh**
+
+```shell
+kustomize build . --enable-helm > /dev/null
+
+get_parameters() {
+while read -r chart; do  
+  yq e -o=p "charts/$chart/values.yaml" | jq --arg chart "$chart" -nR 'inputs | sub(" .*"; "") | {name: ., section: "\($chart) Helm chart properties"}'
+done << EOF
+$(yq e '.helmCharts[].name' kustomization.yaml)
+EOF
+}
+
+get_parameters | jq --slurp
+```
+
+**Dockerfile**
+
+```dockerfile
+FROM ubuntu:20.04
+
+RUN apt install jq yq helm kustomize -y
+
+ADD get-parameters.sh /home/argocd/get-parameters.sh
+```
+
+#### Example 3: simple Helm CMP
+
+This example demonstrates how the Helm parameters interface could be achieved with a parameterized CMP.
+
+![Helm parameters interface](images/helm-parameters.png)
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: ConfigManagementPlugin
+metadata:
+  name: simple-helm-cmp
+spec:
+  version: v1.0
+  generate:
+    command: [/home/argocd/generate.sh]
+  discover:
+    fileName: "./values.yaml"
+  parameters:
+    command: [/home/argocd/get-parameters.sh]
+
+```
+
+**generate.sh**
+
+```shell
+VALUES_FILES=$(echo "$ARGOCD_PARAMETERS" | jq -r '.["main"][] | select(.name == "values files").value')
+# Put the extra values at a random filename to avoid conflicting with existing files.
+EXTRA_VALUES_FILENAME="values-$(openssl rand -base64 12).yaml"
+echo "$ARGOCD_PARAMETERS" | jq -r '.["main"][] | select(.name == "values").value' > "$EXTRA_VALUES_FILENAME"
+# Convert JSON parameters to comma-delimited k=v pairs.
+PARAMETERS=$(echo "$ARGOCD_PARAMETERS" | jq -r '.["parameters"] | map("\(.name)=\(.value)") | join(",")')
+helm template --values "$VALUES_FILES" --values "$EXTRA_VALUES_FILENAME" --set "$PARAMETERS"
+```
+
+**get-parameters.sh**
+
+```shell
+# Pull params from values.yaml and then append the default ("main") parameters announcements.
+yq e -o=p values.yaml | jq -nR 'inputs | sub(" .*"; "") | {name: ., section: "Properties"}' |
+jq --slurp '[{"name": "values files", "uiConfig": "{\"multiline\": true}"}, {"name": "values"}] + .'
+```
+
+### Security Considerations
+
```

Comment:
This section is noticeably absent. Since we are executing user-defined commands (operator-defined ones which are possibly affected by end-user input), we need to understand and document the risk here. A poorly written plugin could cause security issues (e.g. one that does not escape strings properly).

Reply:
@jessesuen I structured the document a bit weirdly. That information is here, just one more heading down. :-)

If we add another `####` section under security, I'll keep the `Increased scripting` heading. Otherwise I'll drop it to avoid confusion.